### DISCUSSION TITLE
Incompatible props when using with react-timezone-select

### DISCUSSION THREAD

**Discussion by codeforcesmeme** [csandman/chakra-react-select#252] [Q&A]
Hi, I've been using this wrapper with a system that uses chakra-ui and it's been amazing, thank you for maintaining this.

Recently we wanted to integrate this https://github.com/ndom91/react-timezone-select component and thought I could use the hook `useChakraSelectProps` to seamlessly integrate the two but there were a lot of type conflicts, why is that, is it something on my part to define/edit?

I created a small reproducible example here: [https://codesandbox.io/s/react-timezone-chakra-select-h6p6j0](https://codesandbox.io/s/react-timezone-chakra-select-h6p6j0?file=/src/App.tsx) with a functional one (without chakra-react-select) is in `App-default.tsx`.

The type errors are mainly two, one in the `onChange` of the hook and others are `aria-errormessage` `aria-invalid` and some others.
I'm not entirely sure whether the errors are due types/generics shenanigans in either chakra-react-select or react-timezone-select but thought to post here and maybe mention @ndom91 as well.

**Comment by codeforcesmeme** [comment]
**Update:**

I did solve the problem(s) looking at the codebase of both packages I added the types `unknown` and `boolean` to the `useChakraSelectProps` hook to be compatible with `react-timezone-select` and faced an error in the hook with the `onChange` argument so I had to override in the component props.

Complete example (for documentation, if I remove the codesandbox in the future)
```tsx
import { useState } from "react";
import TimezoneSelect, { ITimezone } from "react-timezone-select";
import { useChakraSelectProps } from "chakra-react-select";

import "./styles.css";
export default function App() {
  const [timezone, setTimezone] = useState<ITimezone>(
    Intl.DateTimeFormat().resolvedOptions().timeZone
  );
  const selectProps = useChakraSelectProps<unknown, boolean>({
    value: timezone,
    onChange: () => {}
  });

  return (
    <div className="App">
      <h2>
        <code>react-timezone-select</code> + <code>chakra-react-select</code>
      </h2>
      <div className="timezone--wrapper">
        <TimezoneSelect
          {...selectProps}
          value={timezone}
          onChange={setTimezone}
        />
      </div>
    </div>
  );
}
```

**Reply by csandman** [reply]
I actually ended up taking a look at this before I noticed that you posted an update. The problem is actually due to generic shenanigans and from what I can tell, it's the fault of `react-timezone-select`.  Realistically, they should have generics on their extension of the props interface they're extending from `react-select`, because they do have a strict typing for `Option` and `IsMulti`.

I actually made a version of their package implementing the generics to see how it worked, and it worked better for me. I may end up submitting a PR to fix it haha.

---

On a side note, when I made the `useChakraSelectProps` hook, I didn't really think about what it needed in order to work right. Ultimately, I've realized it doesn't actually need the `onChange` or `value` props forwarded through it. Or many other props for that matter. The only ones it really needs are ones the ones it interacts with inside the hook, the rest it just forwards to the output.  You can see which ones need to be passed into it here: https://github.com/csandman/chakra-react-select/blob/546e49e110b809a9705482fe583df657b96c158b/src/use-chakra-select-props.ts

What I'm getting at, is that keeping the `value` and `onChange` props entirely separated from it is an ok strategy. In fact, you can call `useChakraSelectProps` with an empty object and it should still have the same result:

```ts
const selectProps = useChakraSelectProps<unknown, boolean>({});
```

In fact, I should probably make the props argument for the hook optional, because being forced to pass an empty object is dumb haha.

I also might make it so only the necessary props are allowed in the hook, but I'll have to think about it more. I mostly just exported the hook as an afterthought, because I thought, "who knows, maybe someone will get some use out of it" haha.

**Reply by codeforcesmeme** [reply]
Ah yea, makes sense. I saw your fork https://github.com/csandman/react-timezone-select-fork/commit/63f6993ac8c8f0f480c795002e1c31e186715784 and the changes make sense! Would love if you can make a PR @ the react-timezone-select and it be merged to avoid workarounds and pain points for future users.

Also I very much admire what you've done with the hook, making use cases like mine possible! 
The empty object can be avoided yea -but it's a big deal really-.
Thank you very much for your help and engagement!

**Reply by csandman** [reply]
> Ah yea, makes sense. I saw your fork [csandman/react-timezone-select-fork@63f6993](https://github.com/csandman/react-timezone-select-fork/commit/63f6993ac8c8f0f480c795002e1c31e186715784) and the changes make sense! Would love if you can make a PR @ the react-timezone-select and it be merged to avoid workarounds and pain points for future users.

I'll probably do that! I've been insanely busy these past couple weeks as I just put my two weeks in at my current job, and they're trying to get everything they possibly can out of me haha. But I'm off next week, so I'll take a look at it then.

> Also I very much admire what you've done with the hook, making use cases like mine possible!

I'm glad you like it! It's funny, there's actually a request for a similar feature on `react-timezone-select` haha: https://github.com/ndom91/react-timezone-select/issues/64

I added it as an afterthought (because of `react-google-places-autocomplete` specifically) so I'm glad at least one person is getting some use out of it!

**Reply by ndom91** [reply]
Hey thanks for the deep dive everyone! If typescript / generics are the issue here, I'd love a PR - I'm no TS expert :pray: 

Nevermind, its just that one commit? (https://github.com/csandman/react-timezone-select-fork/commit/63f6993ac8c8f0f480c795002e1c31e186715784)

I can add it today :+1:

**Reply by ndom91** [reply]
PR is up: https://github.com/ndom91/react-timezone-select/pull/89

**Reply by ndom91** [reply]
Should be fixed in `1.4.1+`

Thanks for the suggested fix @csandman :pray:

**Reply by csandman** [reply]
@ndom91 Thanks for hopping in the discussion! Yes, it is just that one commit. The only reason it's really necessary in this case is because the props object is being defined outside the select component. Normally, the component will infer the option types based on what you pass into it. But by making a wrapper, you either have to accept the generics into the component, or define them yourself. And because in this case, the typing of the option objects is known, passing them yourself is the easiest way to fix this!

**Reply by ndom91** [reply]
Okay great, I pushed a few other open PRs to the library last night / today. 

If you could check out `1.5.4+` and let me know if this use-case works for you, that's be great 🙏

**Reply by csandman** [reply]
So I'm having trouble with `v1.5.4` on CodeSandbox, for some reason it's giving me an export error: https://codesandbox.io/s/chakra-react-timezone-select-yx8hl4?file=/app.tsx
